enterprise/coordinator/recovery/recovery_test.go

```diff
@@ -0,0 +1,272 @@
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/goleak"
+	"google.golang.org/grpc"
+	testingclock "k8s.io/utils/clock/testing"
+)
+
+func TestPeriodically(t *testing.T) {
+	require := require.New(t)
+	ctx, cancel := context.WithCancel(t.Context())
+	errs := make(chan error, 1)
+
+	clock := testingclock.NewFakeClock(time.Now())
+	interval := time.Second
+	calls := make(chan struct{}, 15)
+
+	go func() {
+		errs <- periodically(ctx, clock, interval, func() {
+			calls <- struct{}{}
+		})
+	}()
+
+	for range 12 {
+		// The function should be called once per clock step.
+		receiveEventually(t, time.Millisecond, calls)
+		clock.Step(interval)
+	}
+
+	cancel()
+	err := receiveEventually(t, time.Second, errs)
+	require.ErrorIs(err, context.Canceled)
+}
+
+func receiveEventually[A any](t *testing.T, d time.Duration, ch chan A) A {
+	// t.FailNow, require.Fail etc. are not considered function returns, so the compiler complains
+	// about a missing return after the select. Wrapping the select in a for loop makes the error
+	// disappear.
+	for {
+		select {
+		case <-time.After(d):
+			t.Fatalf("no object received within %s", d.String())
+		case v := <-ch:
+			return v
+		}
+	}
+}
+
+func TestRecoverOnce(t *testing.T) {
+	logger := slog.Default()
+	ctx := t.Context()
+
+	for name, tc := range map[string]struct {
+		peerGetter peerGetter
+		guard      guard
+		dial       func(_ context.Context, issuer atls.Issuer, validators []atls.Validator, _ *slog.Logger, addr string) (meshapi.MeshAPIClient, func() error, error)
+		wantErr    error
+	}{
+		"no peers": {
+			peerGetter: &fakePeerGetter{nil, nil},
+			guard:      newStaleGuard(t),
+			dial: func(context.Context, atls.Issuer, []atls.Validator, *slog.Logger, string) (meshapi.MeshAPIClient, func() error, error) {
+				return newFakeClient(t), func() error { return nil }, nil
+			},
+			wantErr: errNoPeers,
+		},
+		"bad peerGetter": {
+			peerGetter: &fakePeerGetter{nil, assert.AnError},
+			guard:      newStaleGuard(t),
+			dial: func(context.Context, atls.Issuer, []atls.Validator, *slog.Logger, string) (meshapi.MeshAPIClient, func() error, error) {
+				return newFakeClient(t), func() error { return nil }, nil
+			},
+			wantErr: assert.AnError,
+		},
+		"bad dial": {
+			peerGetter: &fakePeerGetter{[]string{"foo"}, nil},
+			guard:      newStaleGuard(t),
+			dial: func(context.Context, atls.Issuer, []atls.Validator, *slog.Logger, string) (meshapi.MeshAPIClient, func() error, error) {
+				return nil, nil, assert.AnError
+			},
+			wantErr: assert.AnError,
+		},
+		"one bad peer": {
+			peerGetter: &fakePeerGetter{[]string{"a", "b"}, nil},
+			guard:      newStaleGuard(t),
+			dial: func(_ context.Context, _ atls.Issuer, _ []atls.Validator, _ *slog.Logger, addr string) (meshapi.MeshAPIClient, func() error, error) {
+				if addr == "b:7777" {
+					return newFakeClient(t), func() error { return nil }, nil
+				}
+				return nil, nil, assert.AnError
+			},
```

Comment:
hard to read, a stubDialer could have a map from addr to result.

Reply:
Done